plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/preferences/RunOptionsDefaultsComponent.java

```diff
@@ -0,0 +1,312 @@
+    // We can't retrieve staging locations because no project was input or we're not authenticated.
+    if (!Strings.isNullOrEmpty(project) && client != null) {
+      ListenableFutureProxy<SortedSet<String>> stagingLocationsFuture =
+          FetchStagingLocationsJob.schedule(client, project);
+      UpdateStagingLocationComboListener updateComboListener =
+          new UpdateStagingLocationComboListener(stagingLocationsFuture);
+      stagingLocationsFuture.addListener(updateComboListener, executor);
+    }
+  }
+
+  /**
+   * Ensure the staging location specified in the input combo is valid.
+   */
+  private void verifyStagingLocation(final String stagingLocation) {
+    if (verifyJob != null) {
+      // Cancel any existing verifyJob
+      verifyJob.cancel();
+    }
+    if (client == null) {
+      // We can't verify the staging locations because we don't have a GCS client
+      return;
+    }
+    if (stagingLocation.isEmpty()) {
+      // If the staging location is empty, we don't have anything to verify; and we don't have any
+      // interesting messaging
+      messageTarget.clear();
+      return;
+    }
+    verifyJob = VerifyStagingLocationJob.create(client, stagingLocation);
+    verifyJob.schedule(VERIFY_LOCATION_DELAY_MS);
+    final ListenableFutureProxy<VerifyStagingLocationResult> resultFuture =
+        verifyJob.getVerifyResult();
+    resultFuture.addListener(
+        new Runnable() {
+          @Override
+          public void run() {
+            try {
+              VerifyStagingLocationResult result = resultFuture.get();
+              if (result.getStagingLocation().equals(stagingLocationInput.getText())
+                  && result.isAccessible()) {
+                messageTarget.setInfo("Found staging location " + stagingLocation);
+              } else {
+                messageTarget.setError(String.format("Couldn't fetch bucket %s", stagingLocation));
+              }
+            } catch (InterruptedException | ExecutionException e) {
+              messageTarget.setError(String.format("Couldn't fetch bucket %s", stagingLocation));
+            }
+          }
+        },
+        executor);
+  }
+
+  /**
+   * Whenever focus is lost, retrieve all of the buckets and update the target combo with the
+   * retrieved buckets, and update the {@link SelectFirstMatchingPrefixListener} with new
+   * autocompletions.
+   */
+  private class GetProjectStagingLocationsListener extends FocusAdapter {
+    @Override
+    public void focusLost(FocusEvent e) {
+      updateStagingLocations(getProject());
+    }
+  }
+
+  /**
+   * Create a GCS bucket in the project specified in the project input at the location specified in
+   * the staging location input.
+   */
+  private class CreateStagingLocationListener extends SelectionAdapter {
+    @Override
+    public void widgetSelected(SelectionEvent e) {
+      if (client == null) {
+        return;
+      }
+      String projectName = projectInput.getText();
```

Comment:
There's a place in the App Engine code where we could use this. (We throw an exception if a requested staging bucket doesn't already exist.) Just making a note to myself to remember this.